### PR TITLE
 perf: Use the minimum integer size necessary to represent states 

### PR DESCRIPTION
**Pull request by Marwes**
Part of #304 . We detect how many reductions and states that is needed before generating any code and then use the minimum necessary integer size to hold all of these. Most grammars likely fit within `i16` but smaller ones may fit within `i8`
.
The first commit simplifies the generated code a bit but did not help reduce the binary code in any way. Might make it a bit less work for LLVM at least.

### CODE REVIEW

File: lalrpop/src/lr1/codegen/parse_table.rs [hunk 1]
```diff
@@ -332,6 +334,16 @@ impl<'ascent, 'grammar, W: Write> CodeGenerator<'ascent, 'grammar, W, TableDrive
                                                                       .zip(0..)
                                                                       .collect();
 
+        let state_type = {
+            if states.len() <= ::std::i8::MAX as usize && reduce_indices.len() <= (::std::i8::MAX as usize + 1) {
```
Comment (nikomatsakis) [inline lalrpop/src/lr1/codegen/parse_table.rs]: Nit: can we do:

```rust
let max_value = ::std::cmp::max(states.len(), reduce_indices.len() - 1);
if max_value <= ::std::i8::MAX { ... }
```

also, worth commenting where the `-1` comes from =)

**Comment by ahmedcharles** [comment]
My only nit is that making the travis changes independently may be beneficial.

**Comment by nikomatsakis** [comment]
> My only nit is that making the travis changes independently may be beneficial.

true...

File: lalrpop/src/lr1/codegen/parse_table.rs [hunk 2]
```diff
@@ -332,6 +334,19 @@ impl<'ascent, 'grammar, W: Write> CodeGenerator<'ascent, 'grammar, W, TableDrive
                                                                       .zip(0..)
                                                                       .collect();
 
+        let state_type = {
+            // `reduce_indices` are allowed to be +1 since the negative maximum of any integer type
+            // is one larger than the positive maximum
+            let max_value = states.len().max(reduce_indices.len() - 1);
```
Comment (Dylan-DPC) [inline lalrpop/src/lr1/codegen/parse_table.rs]: you are using unstable library feature 'ord_max_min'

Comment (Marwes) [reply lalrpop/src/lr1/codegen/parse_table.rs]: Oops, didn't have time to run through the full tests, was sure that I had used this on stable already.

**Pull request merged**